### PR TITLE
[youtube] Extract data from multiple clients

### PR DESCRIPTION
**Pull request by pukkandan**
Review and more testing needed
@colethedj 

TODO:
* [x] documentation of the new player_client choices
* [x] Test that it closes https://github.com/yt-dlp/yt-dlp/issues/501
* [x] Test impact on videos mentioned in https://github.com/yt-dlp/yt-dlp/issues/502

### CODE REVIEW

File: yt_dlp/extractor/youtube.py [hunk 1]
```diff
@@ -2287,228 +2302,122 @@ def _get_video_info_params(video_id, client='TVHTML5'):
...
-                        item_id=video_id, ep='player', query=yt_age_query,
-                        ytcfg=ytcfg_age, headers=ytage_headers, fatal=False,
-                        default_client=yt_client,
-                        note='Downloading %sage-gated player API JSON' % ('android ' if force_mobile_client else '')
-                    ) or {}
+        session_index = self._extract_session_index(ytpcfg, master_ytcfg)
+        syncid = self._extract_account_syncid(ytpcfg, master_ytcfg, initial_pr)
+        sts = self._extract_signature_timestamp(video_id, player_url, master_ytcfg, fatal=False)
+        headers = self._generate_api_headers(
+            master_ytcfg, identity_token, syncid, client=self._YT_CLIENTS[client], session_index=session_index)
```
Comment (colethedj) [inline yt_dlp/extractor/youtube.py:2311]:
```suggestion
        headers = self._generate_api_headers(
            ytpcfg, identity_token, syncid, client=self._YT_CLIENTS[client], session_index=session_index)
```

should be ytpcfg so we don't accidentally generate the web client's headers if something goes wrong

File: yt_dlp/extractor/youtube.py [hunk 2]
```diff
@@ -2287,228 +2302,122 @@ def _get_video_info_params(video_id, client='TVHTML5'):
...
+                if client == 'web_music' and 'configs' not in self._configuration_arg('player_skip'):
+                    ytm_webpage = self._download_webpage(
+                        'https://music.youtube.com',
+                        video_id, fatal=False, note='Downloading remix client config')
+                    ytpcfg = self._extract_ytcfg(video_id, ytm_webpage) or {}
+                pr = self._extract_player_response(
+                    client, video_id, ytpcfg or master_ytcfg, ytpcfg, identity_token, player_url, initial_pr)
+            if pr:
+                yield pr
+            if traverse_obj(pr, ('playabilityStatus', 'reason')) in self._AGE_GATE_REASONS:
```
Comment (colethedj) [inline yt_dlp/extractor/youtube.py:2396]: we should make note of when we have detected the video is age-gated so we don't bother extracting any more non-age gate pr's from additional clients.

File: yt_dlp/extractor/youtube.py [hunk 3]
```diff
@@ -2387,18 +2390,20 @@ def _extract_player_responses(self, clients, video_id, webpage, ytcfg, player_ur
...
-                    client, video_id, ytpcfg or ytcfg, identity_token, player_url, initial_pr)
+                    client, video_id, ytpcfg or master_ytcfg, identity_token, player_url, initial_pr)
                 if pr:
                     yield pr
         # Android player_response does not have microFormats which are needed for
         # extraction of some data. So we return the initial_pr with formats
         # stripped out even if not requested by the user
         # See: https://github.com/yt-dlp/yt-dlp/issues/501
+
+        # TODO: think there is adaptive formats
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2406]: formats, adaptiveFormats, hlsManifestUrl, dashManifestUrl are all inside streamingData

File: yt_dlp/extractor/youtube.py [hunk 4]
```diff
@@ -2834,9 +2837,13 @@ def process_language(container, base_url, lang_code, sub_name, query):
                 webpage, self._YT_INITIAL_DATA_RE, video_id,
                 'yt initial data')
         if not initial_data:
+            headers = self._generate_api_headers(
+                master_ytcfg, identity_token, self._extract_account_syncid(master_ytcfg),
+                session_index=self._extract_session_index(master_ytcfg))
+
```
Comment (pukkandan) [inline yt_dlp/extractor/youtube.py:2843]: header are already defined at the top of `_real_extract`

File: yt_dlp/extractor/youtube.py [hunk 5]
```diff
@@ -2287,228 +2302,122 @@ def _get_video_info_params(video_id, client='TVHTML5'):
...
-                        })
-                        feed_ids.append(feed_id)
-                    self.to_screen(
-                        'Downloading multifeed video (%s) - add --no-playlist to just download video %s'
-                        % (', '.join(feed_ids), video_id))
-                    return self.playlist_result(
-                        entries, video_id, video_title, video_description)
+        for client in clients:
+            ytpcfg = master_ytcfg if client == 'web' else {}
+            if client == 'web' and initial_pr:
```
Comment (colethedj) [inline yt_dlp/extractor/youtube.py:2384]: `initial_pr` may not be defined?

Comment (colethedj) [reply yt_dlp/extractor/youtube.py:2384]: nvm

**Pull request merged**